Road/source/features/data_cleaning.py

```diff
@@ -0,0 +1,168 @@
+
+def convert_emoticons(text: str) -> str:
+    """
+    Convert emoticons in text to their corresponding text representation.
+
+    Args:
+        text (str): The input text with emoticons.
+
+    Returns:
+        str: The input text with emoticons replaced by their corresponding text representation.
+    """
+    emo_obj = emot.EMOTICONS_EMO
+    for emot1 in emo_obj:
+        escaped_emot = re.escape(emot1)
+        text = re.sub(u'({})'.format(escaped_emot), "_".join(emo_obj[emot1].replace(",","").split()), text)
+    return text
+
+def remove_urls(text: str) -> str:
+    """
+    Remove URLs from text.
+
+    Args:
+        text (str): The input text with URLs.
+
+    Returns:
+        str: The input text with URLs removed.
+    """
+    url_pattern = re.compile(r'https?://\S+|www\.\S+')
+    return url_pattern.sub(r'', text)
+
+def remove_phone_numbers(text: str) -> str:
+    """
+    Remove phone numbers from text.
+
+    Args:
+        text (str): The input text with phone numbers.
+
+    Returns:
+        str: The input text with phone numbers removed.
+    """
+    text = re.sub(r'(\+?\d{2,4}[ -]?)?\d{9,10}', '', text)
+    return text
+
+PUNCT_TO_REMOVE = string.punctuation + '؛،؟«»٪٫٬٭'
+def remove_punctuations(text: str) -> str:
+     """
+     Remove punctuations from text.
+
+     Args:
+          text (str): The input text with punctuations.
+
+     Returns:
+          str: The input text with punctuations removed.
+     """
+     return text.translate(str.maketrans('', '', PUNCT_TO_REMOVE))
+
+def remove_multimedia(text: str) -> str:
+    """
+    Remove multimedia (images, videos, etc.) from text.
+
+    Args:
+        text (str): The input text with multimedia.
+
+    Returns:
+        str: The input text with multimedia removed.
+    """
+    text = re.sub(r'http\S+', '', text)
+    text = re.sub(r'pic\.twitter\.com/\S+', '', text)
+    text = re.sub(r'@(\w+)', '', text)
+    return text
+
+def map_reply_messages(data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Map each message to its corresponding reply message in the "reply" column, and delete any rows with non -1 reply_to values.
+
+    Args:
+        data (pandas.DataFrame): The input DataFrame containing message data.
+
+    Returns:
+        pandas.DataFrame: The modified DataFrame with mapped reply messages and deleted rows.
+    """
+    # create an empty dictionary to store message-reply mappings
+    message_reply_dict = {}
+
+    # iterate over each row in the dataframe
+    for i, row in data.iterrows():
+        # check if the row has a valid reply_to value
+        if row["reply_to"] != -1:
+            # store the message-reply mapping in the dictionary
+            message_reply_dict[row["reply_to"]] = row["message"]
+
+    # update the reply column with the corresponding reply message
+    data["reply"] = data["id"].map(message_reply_dict)
+
+    # delete any rows with non -1 reply_to values
+    data = data[data["reply_to"] == -1]
```

Comment:
this filter the dataframe to be the value of column -1, no need for the column after filtered.